roles/internetarchive/tasks/main.yml

```diff
@@ -1,32 +1,93 @@
-# We need a recent version of node
-- name: FAIL (STOP INSTALLING) IF nodejs_version is not set to 10.x or 12.x
-  fail:
-    msg: "Internet Archive install cannot proceed, as it currently requires Node.js 10.x or 12.x, and your nodejs_version is set to {{ nodejs_version }}.  Please check the value of nodejs_version in /opt/iiab/iiab/vars/default_vars.yml and possibly also /etc/iiab/local_vars.yml"
-  when: internetarchive_install and (nodejs_version != "10.x") and (nodejs_version != "12.x")
+# "How do i fail a task in Ansible if the variable contains a boolean value?
+# I want to perform input validation for Ansible playbooks"
+# https://stackoverflow.com/questions/46664127/how-do-i-fail-a-task-in-ansible-if-the-variable-contains-a-boolean-value-i-want/46667499#46667499
 
-- name: Install Yarn and Internet Archive
+# We assume 0-init/tasks/validate_vars.yml has DEFINITELY been run, so no need
+# to re-check whether vars are defined here.  As Ansible vars cannot be unset:
+# https://serverfault.com/questions/856729/how-to-destroy-delete-unset-a-variable-value-in-ansible
+
+- name: Assert that "internetarchive_install is sameas true" (boolean not string etc)
+  assert:
+    that: internetarchive_install is sameas true
+    fail_msg: "PLEASE SET 'internetarchive_install: True' e.g. IN: /etc/iiab/local_vars.yml"
+    quiet: yes
+
+- name: Assert that "internetarchive_enabled | type_debug == 'bool'" (boolean not string etc)
+  assert:
+    that: internetarchive_enabled | type_debug == 'bool'
+    fail_msg: "PLEASE GIVE VARIABLE 'internetarchive_enabled' A PROPER (UNQUOTED) ANSIBLE BOOLEAN VALUE e.g. IN: /etc/iiab/local_vars.yml"
+    quiet: yes
+
+
+#- name: Set --reinstall fact
+#  set_fact:
+#    internetarchive_upgrade: True
+#  when: reinstall is defined
+
+- block:    # UPDATE IF... internetarchive_installed is defined and internet_available
+
+  - name: Stop 'internetarchive' systemd service, if internetarchive_upgrade
+    systemd:
+      name: internetarchive
+      daemon_reload: yes
+      state: stopped
+    #when: internetarchive_enabled and internetarchive_upgrade
+
+  - name: Update pre-existing install (yarn upgrade) if internetarchive_upgrade
+    shell: yarn config set child-concurrency 1 && yarn install && yarn upgrade
+    args:
+      chdir: "{{ internetarchive_dir }}"
+    #when: internetarchive_enabled and internetarchive_upgrade
+
+  when: internetarchive_installed is defined and internet_available   # END BLOCK
```

Comment:
Painful always running upgrade when ansible runs, and risky as unplanned changes could be introduced like what occurred in #2204.

Reply:
Excellent point.  @mitra42 and I should discuss Community QA strategy when we meet in person here in Arizona in a few hours.

"Customers" who aren't paying should definitely help with QA (paying it forward) but at same time we/they/all need to bracket & mitigate the worst risks in creative ways.

So this is a really great discussion point towards tuning up near-fieldback not just distant-fieldback...

Reply:
Definitely not the final word, but I'd like to merge now, while at the same time let's keep thinking thru these `--reinstall` and `--upgrade` possibilities to get this right in 2020.

Reply:
I'll go with this role is the only role forcing an upgrade to be preformed post-install deviating from the behavior of all the other roles. This results in a large increase(50%) in post-install runtime of ICO or iiab-configure was ~9mins on large, now at ~15mins. footnote while testing #2200 which had a runtime of 6.5mins or less and cracked 6min a couple of times so this is going the wrong way IMHO.

Reply:
> this role is the only role forcing an upgrade to be preformed post-install

Many roles traditionally have upgraded after installation, for example WordPress, and for very good reasons — even if deciding when security is a priority, when ongoing community dogfooding (community testing) is a priority, when databases should be wiped completely clean — are tough decisions where the answer will not always be the same for each app.

Etc.  Among other post-install desires, which are the reason that flags like `--upgrade` are not at all an easy topic — but some/such will eventually be needed!

Reply:
Just to restate what was said earlier on this tkt and others — internetarchive and/or others will likely soon want their Ansible roles to run into 3-or-4-or-5-or-more different ways, e.g.:
1) `--reinstall` while preserving all of the app's db's, content & metrics/logs
2) `--reinstall` after WIPING all app's db's, content (and ideally wiping all of its associated privacy/analytics/metrics/logs)
3) `--upgrade` an app
4) Disable-or-enable an already installed app
5) Other scenarios that will surely emerge...both online and offline

Reply:
I think this is an important conversation that shouldnt be buried in a comment on a MR that is already merged :-) 

I agree with @jvonau that we dont want an automatic update of Internet Archive's code to slow down other completely unrelated tasks. 

I also agree with @holta that we should think about what flags are needed so there is a template and consistent way to express different requirements. I'm not quit sure of the nuances of the differences between his options 1 and 2 above, but I certainly see three completely difference scenarios 
a) Upgrade - make sure the box is running the latest code, this is particularly important when bugs are getting fixed rapidly (Wordpress is a good example of this). 
b) Reinstall - something went wrong, lets clean up as much as we can and try again. Typically I'm going to want to preserve as much data as possible in that case. 
c) Deinstall - take the app off, and wipe the data. 

And especially "a" and "b" need to be one-liners, since its the kind of thing I'd expect to be doing via remote support, when I certainly dont want to be telling someone to go and edit this variable, or delete this directory etc etc. 

One impressive thing on the OLIP demo was the way it was easy to add / delete / update apps from a UI.

Reply:
> c) Deinstall - take the app off, and wipe the data.

Yes, we might start experimenting with this earlier than I expected.  IIAB file /etc/iiab/iiab_state.yml now makes this opportunity possible.

> One impressive thing on the OLIP demo was the way it was easy to add / delete / update apps from a UI.

Yes.  That was one aspect of OLIP that appeared to work extremely well — somewhat similar to Kolibri — offering a very clean and simple GUI that just works as people expect (requiring extensive community listening to refine & get right!) is becoming critical in this new decade...

Reply:
Thanks to @mitra42 for calling today, further evolving this preliminary design discussion, around what folks truly need from ./runrole and similar post-install scenarios.  He explained that internetarchive requires only 2 code paths for now (for IIAB 7.1) and that is more than sufficient for now.

But that we all agree to think about 3 code paths e.g. something like `./runrole --upgrade internetarchive` in future if that proves necessary.

As summarized below, there is no significant delay to running `yarn upgrade` every time an implement is online (as documented at PR #2248) and regardless even if there was, any larger delay would be most worth it at this phase in [internetarchive](https://github.com/iiab/iiab/tree/master/roles/internetarchive/)'s life cycle &mdash; as this increases testing of the latest internetarchive towards it becoming a more mature product:  https://github.com/iiab/iiab/blob/master/roles/internetarchive/tasks/main.yml#L22-L30